alg_convex_hull.go

```diff
@@ -1,110 +1,191 @@
-	rightTurn = iota
+	rightTurn threePointOrientation = iota + 1
 	// collinear indicates three points are on the same line
 	collinear
 	// leftTurn indicates the orientation is left turn which is clockwise
 	leftTurn
 )
 
-// graphamScan returns a convex hull.
+func (o threePointOrientation) String() string {
+	switch o {
+	case rightTurn:
+		return "right turn"
+	case collinear:
+		return "collinear"
+	case leftTurn:
+		return "left turn"
+	default:
+		return "invalid orientation"
+	}
+}
+
+// orientation checks if s is on the right hand side or left hand side of the line formed by p and q.
+func orientation(p, q, s XY) threePointOrientation {
+	cp := q.Sub(p).Cross(s.Sub(q))
+	switch {
+	case cp.GT(zero):
+		return leftTurn
+	case cp.LT(zero):
+		return rightTurn
+	default:
+		return collinear
+	}
+}
+
+func convexHull(g Geometry) Geometry {
+	if g.IsEmpty() {
+		// special case to mirror postgis behaviour
+		return g
+	}
+	pts := g.convexHullPointSet()
+	hull := grahamScan(pts)
+	switch len(hull) {
+	case 0:
+		return NewGeometryCollection(nil)
+	case 1:
+		return NewPoint(hull[0])
+	case 2:
+		ln, err := NewLine(
+			Coordinates{hull[0]},
+			Coordinates{hull[1]},
+		)
+		if err != nil {
+			panic("bug in grahamScan routine - output 2 coincident points")
+		}
+		return ln
+	default:
+		coords := [][]Coordinates{make([]Coordinates, len(hull))}
+		for i := range hull {
+			coords[0][i] = Coordinates{XY: hull[i]}
+		}
+		poly, err := NewPolygonFromCoords(coords)
+		if err != nil {
+			panic(fmt.Errorf("bug in grahamScan routine - didn't produce a valid polygon: %v", err))
+		}
+		return poly
+	}
+}
+
+type pointStack []XY
+
+func (s *pointStack) push(p XY) {
+	(*s) = append(*s, p)
+}
+
+func (s *pointStack) pop() XY {
+	p := s.top()
+	(*s) = (*s)[:len(*s)-1]
+	return p
+}
+
+func (s *pointStack) top() XY {
+	return (*s)[len(*s)-1]
+}
+
+func (s *pointStack) underTop() XY {
+	return (*s)[len(*s)-2]
+}
+
+// grahamScan returns the convex hull of the input points. It will either
+// represent the empty set (zero points), a point (one point), a line (2
+// points), or a closed polygon (>= 3 points).
 func grahamScan(ps []XY) []XY {
-	if len(ps) < 3 {
-		return nil
+	if len(ps) <= 1 {
+		return ps
 	}
 
 	sortByPolarAngle(ps)
 
-	resultStack := make([]XY, 2, len(ps))
-	copy(resultStack, ps[:2])
-	toDoStack := make([]XY, len(ps)-2)
-	copy(toDoStack, ps[2:])
+	// Append the lowest-then-leftmost point so that the polygon will be closed.
+	ps = append(ps, ps[0])
 
-	for i := 0; i < len(toDoStack); i++ {
-		ori := orientation(resultStack[len(resultStack)-2], resultStack[len(resultStack)-1], toDoStack[i])
-		switch {
-		case ori == leftTurn:
-			resultStack = append(resultStack, toDoStack[i])
-		default:
-			resultStack = resultStack[:len(resultStack)-1]
-			resultStack = append(resultStack, toDoStack[i])
+	var i int // tracks progress through the ps slice
+	var stack pointStack
+	stack.push(ps[0])
+	i++
+	for i < len(ps) && len(stack) < 2 {
```

Comment:
why do you not simply push the first element?

Reply:
The first element is pushed on line 114. But then we need the `for` loop to account for the cases where there are duplicated points equal to the first point. E.g. if the points are `(3 4), (3 4), (5 5), ...`, we push point `(3 4)` on line 114, then the for loop skips over the second point (also `(3 4`), and then pushes the 3rd point `(5 5)`.

Reply:
Sorry, I meant the second point.

Reply:
The second point may be equal to the first point, in which case we don't want to push it (since we don't want duplicated points in the output). The third point may also be equal to the first, so we need to use a loop to get the first non-conincident point.